### DISCUSSION TITLE
using Rtabmap without UI and in programming dotnet or python

### DISCUSSION THREAD

**Discussion by damavand1** [introlab/rtabmap#1186] [General]
I run Rtabmap using this command 

ros2 launch rtabmap_examples kinect_xbox_360.launch.py

every things fine, I can see yellow and green point's and by moving camera I see 3d map is extending 

But i need some more in programming
1- By programming I want move my robot in environment to scan and save the map
after that, i want save map in memory named (floor 1, floor 2 , ...) and load them every times i need

2- I need define origin point in every 3D  map (X=0,Y=0,Z=0) in every map, and rtabmap detect every time where my robot is in environment and send current location in 3D to my program for example (x=10,Y=0,Z=1000) in cm

My main programming language is C# dot net
but i can do in python

thank you

**Comment by matlabbe** [comment]
If you communicate with rtabmap node with ros topics and services, you could program in any language supported by ros2 to create your own nodes (not sure about c#, but you can program in python for sure).

For teleop or autonomy (nav2 integration), I'll suggest to try turtlebot3 example to see how they did it with turtlebot3: https://github.com/introlab/rtabmap_ros/tree/ros2#example-with-turtlebot3

For managing each floor, you may relaunch rtabmap with different database path, or with a runtime call service `load_database` to load a previously created database, then rtabmap will start localizing in it. To set the origin of the map, the easiest way is to reset the odometry at a particular pose, reset rtabmap, then the first pose rtabmap received will the origin of the new map.

**Comment by damavand1** [comment]
I'm using  

`ros2 launch rtabmap_examples kinect_xbox_360.launch.py`

rtabmap opened and start mapping 
after scanning I saved map in this path:
~/Documents/RTAB-Map/room1.db


now I need a python sample program that load my db from this file and using my Kinect 360 to print on screen current location of camera in environment every moment (x,y,z / r,p,y)

thank you

**Reply by matlabbe** [reply]
If you re-launch in localization mode (remove that [line](https://github.com/introlab/rtabmap_ros/blob/2f365ab7e10025c4d13b7855dfb4b7c8bb01f2d3/rtabmap_examples/launch/kinect_xbox_360.launch.py#L46C20-L46C30) and add to [parameters](https://github.com/introlab/rtabmap_ros/blob/2f365ab7e10025c4d13b7855dfb4b7c8bb01f2d3/rtabmap_examples/launch/kinect_xbox_360.launch.py#L11-L16) `'Mem/IncrementalMemory': 'True'`, note that you can also add `'database_path':'~/Documents/RTAB-Map/room1.db'`), you can first test this:
```
ros2 run tf2_ros tf2_echo map camera_link
```
to see the pose of the camera in map frame (note that the camera should localize first, point it to a known area of the map with similar point of view).

With a python node, you could use a [TransformListener](https://docs.ros.org/en/foxy/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Py.html) and get the same transform in a loop.
